swarm/storage/netstore.go

```diff
@@ -18,318 +18,280 @@ package storage
-		},
-		peers: peers,
-		span:  span,
+		return res, nil
 	}
+
+	ctx, ssp := spancontext.StartSpan(
+		ctx,
+		"localstore.get")
+	defer ssp.Finish()
+
+	return ch, nil
 }
 
-// Fetch fetches the chunk synchronously, it is called by NetStore.Get is the chunk is not available
-// locally.
-func (f *fetcher) Fetch(rctx context.Context) (Chunk, error) {
-	atomic.AddInt32(&f.requestCnt, 1)
-	defer func() {
-		// if all the requests are done the fetcher can be cancelled
-		if atomic.AddInt32(&f.requestCnt, -1) == 0 {
-			f.cancel()
+func (n *NetStore) RemoteFetch(ctx context.Context, req *Request, fi *FetcherItem) error {
+	// while we haven't timed-out, and while we don't have a chunk,
+	// iterate over peers and try to find a chunk
+	metrics.GetOrRegisterCounter("remote.fetch", nil).Inc(1)
+
+	ref := req.Addr
+
+	for {
+		metrics.GetOrRegisterCounter("remote.fetch.inner", nil).Inc(1)
+
+		innerCtx, osp := spancontext.StartSpan(
+			ctx,
+			"remote.fetch")
+		osp.LogFields(olog.String("ref", ref.String()))
+
+		log.Trace("remote.fetch", "ref", ref)
+		currentPeer, err := n.RemoteGet(innerCtx, req, n.localID)
+		if err != nil {
+			log.Error(err.Error(), "ref", ref)
+			osp.LogFields(olog.String("err", err.Error()))
+			osp.Finish()
+			return err
+		}
+		osp.LogFields(olog.String("peer", currentPeer.String()))
+
+		// add peer to the set of peers to skip from now
+		log.Trace("remote.fetch, adding peer to skip", "ref", ref, "peer", currentPeer.String())
+		req.PeersToSkip.Store(currentPeer.String(), time.Now())
+
+		select {
+		case <-fi.Delivered:
+			log.Trace("remote.fetch, chunk delivered", "ref", ref)
+
+			osp.LogFields(olog.Bool("delivered", true))
+			osp.Finish()
+			return nil
+		case <-time.After(timeouts.SearchTimeout):
+			metrics.GetOrRegisterCounter("remote.fetch.timeout.search", nil).Inc(1)
+
+			osp.LogFields(olog.Bool("timeout", true))
+			osp.Finish()
+			break
+		case <-ctx.Done(): // global fetcher timeout
+			log.Trace("remote.fetch, fail", "ref", ref)
+			metrics.GetOrRegisterCounter("remote.fetch.timeout.global", nil).Inc(1)
+
+			osp.LogFields(olog.Bool("fail", true))
+			osp.Finish()
+			return errors.New("chunk couldnt be retrieved from remote nodes")
 		}
-		f.span.Finish()
-	}()
-
-	// The peer asking for the chunk. Store in the shared peers map, but delete after the request
-	// has been delivered
-	peer := rctx.Value("peer")
-	if peer != nil {
-		f.peers.Store(peer, time.Now())
-		defer f.peers.Delete(peer)
 	}
+}
 
-	// If there is a source in the context then it is an offer, otherwise a request
-	sourceIF := rctx.Value("source")
+// Has is the storage layer entry point to query the underlying
+// database to return if it has a chunk or not.
+func (n *NetStore) Has(ctx context.Context, ref Address) (bool, error) {
+	return n.Store.Has(ctx, ref)
+}
 
-	hopCount, _ := rctx.Value("hopcount").(uint8)
+func (n *NetStore) HasWithCallback(ctx context.Context, ref Address, interestedParty string) (bool, *FetcherItem, bool) {
```

Comment:
Right - I will document this.

It `atomically` checks if a chunk is in the localstore, and if it is not returns a `FetcherItem` which will notify the caller upon delivery of the chunk. The name is quite bad, I will see about coming up with a better one.

This function DOES NOT instantiate a network request, but is used to atomically create or return a fetcherItem, so that for a given chunk you have only one fetcherItem.

I agree - the function can be simplified if I split the second part in its own function.

Reply:
Renamed the method to `GetOrCreateFetcherItem` and documented it. Basically the logic is the following:
1. if it returns `ok==true` then you get back a fetcher (created now, or loaded).
2. if it returns `ok==false` then you get a nil fetcher, because the chunk is already in the localstore, and it doesn't make sense to create a fetcher.

I hope that's clear.